Core/Src/main.c

```diff
@@ -87,14 +97,39 @@ int main(void)
   /* Initialize all configured peripherals */
   MX_GPIO_Init();
   MX_USART2_UART_Init();
+  MX_SPI1_Init();
+  MX_TIM1_Init();
   /* USER CODE BEGIN 2 */
 
+  // Set chip select high
+  HAL_GPIO_WritePin(GPIOA, GPIO_PIN_8, GPIO_PIN_SET);
+  HAL_TIM_PWM_Start(&htim1, TIM_CHANNEL_1); 
+
   /* USER CODE END 2 */
 
   /* Infinite loop */
   /* USER CODE BEGIN WHILE */
   while (1)
   {
+    HAL_Delay(10);
+
+    // Get the value from the MCP3004 ADC on channel 0
+    uint8_t spi_tx_data[3] = {MCP3000_START_BIT, MCP3004_CH0, 0x00};
+    uint8_t spi_rx_data[3] = {0};
+    HAL_GPIO_WritePin(GPIOA, GPIO_PIN_8, GPIO_PIN_RESET);
+    HAL_SPI_TransmitReceive(&hspi1, spi_tx_data, spi_rx_data, 3, 100);
+    HAL_GPIO_WritePin(GPIOA, GPIO_PIN_8, GPIO_PIN_SET);
+
+    // The most significant 2 bits are in the second byte, the rest is in the third byte
+    uint16_t adcValue = ((spi_rx_data[1] & 0b00000011) << 8) | spi_rx_data[2];
+    
+    // Calculate the duty cycle
+    float dutyCycle = (float)adcValue / MCP3004_MAX_VALUE;
+    dutyCycle = (dutyCycle * (PWM_MAX_DUTY_CYCLE - PWM_MIN_DUTY_CYCLE)) + PWM_MIN_DUTY_CYCLE;
+    uint16_t pwmCounterValue = (uint16_t)(dutyCycle * PWM_COUNTER_PERIOD);
```

Comment:
So it looks like the calculations check out but it was kinda hard to follow bec of all the naming and reusing variables (specifically "duty cycle"). Please try making the calculations "neater" for readability.

Also remember what duty cycle really means. It is the percentage from 0-100% or 0-1 of the total period...

Reply:
Fixed in f253daa1be4ab5354472ac26b498335741fc8bb2